### ISSUE TITLE
Travsi-ci / any other CI

### ISSUE THREAD

**Issue by davidgraeff** [eclipse/wakaama#182]
Hi,
some time ago I asked for this already (you remember, @dnav :D), but there was no definite answer. Is there any eclipse CI infrastructure that can be integrated into github, to execute the test suite on every commit and for every merge or can a public solution like travis-ci be used?

**Comment by davidgraeff** [comment]
If the maintainers don't want a CI, this issue should be closed.

**Comment by dnav** [comment]
It is more a lack of time than a lack of will. Sorry.

**Comment by rettichschnidi** [comment]
I hacked together a simple PoC using GitHub Actions and SonarCloud: 
* [Repo overview](https://github.com/rettichschnidi/wakaama/tree/rettichschnidi-github-ci)
* [main.yml](https://github.com/rettichschnidi/wakaama/blob/rettichschnidi-github-ci/.github/workflows/main.yml)
* [Resulting report](https://sonarcloud.io/dashboard?id=rettichschnidi_wakaama) (Link will break in the future)

I have not configured SonarCloud, which would be very much needed to make it actually useful. Also, I very much like [Clang Static Analyzer](https://clang-analyzer.llvm.org/) and coverage-reports generated by [lcov](http://ltp.sourceforge.net/coverage/lcov.php). Before adding/rebasing to those tools, I'd like to know the following:

* Would a GitHub runner be acceptable to Eclipse?
* Who would enable SonarCloud access to this Wakaama repository? Maintain the SonarCloud account?

**Comment by sbernard31** [comment]
>    Would a GitHub runner be acceptable to Eclipse?
>    Who would enable SonarCloud access to this Wakaama repository? Maintain the SonarCloud account?

You probably need to ask this to eclipse IT team opening a bugzilla issue : https://bugs.eclipse.org/bugs/enter_bug.cgi?product=Community

**Comment by sbernard31** [comment]
About eclipse CI infrastructure, I think the classic eclipse way is jenkins : https://ci.eclipse.org/
(I don't if there is other way)

**Comment by rettichschnidi** [comment]
I just created an issue in Bugzilla: https://bugs.eclipse.org/bugs/show_bug.cgi?id=568471

**Comment by sbernard31** [comment]
Ok so I think we can use this bugzilla ticket to ask for a JIPP instance for Wakaama project.
But we need first a project lead to be able to give a ["+1 from project lead"](https://wiki.eclipse.org/CBI#Requesting_a_JIPP_instance). (Currently waiting for PMC approval ...)

Next question : 
1. What do you want Jenkins do for you ? (about sonar : https://wiki.eclipse.org/SonarQube)
2. Do you want to use classic config or jenkins pipeline ?

**Comment by rettichschnidi** [comment]
According to [comment 2](https://bugs.eclipse.org/bugs/show_bug.cgi?id=568471#c2) in Bugzilla, it sounds like GitHub is the way to go after all. I do not think that the mentioned services exclusive to Jenkins are needed by Wakaama, and as non-committer, I would not be able to configure Jenkins anyway. Therefore, question number 2 does not seem relevant anymore.

Regarding question 1:
1. Building all executables. Deny PRs if failing.
2. Execute tests. Deny PRs if failing.
3. Measure test coverage. Deny if merging the PR would lower the coverage of the target branch.
4. Integrate static code analysis. Deny PR when new defects detected.

I just created PR #489, which does, if and when merged to master, number 1. It also runs the unit tests, but does not (yet) fail when tests fail. Depending on when PR #423 gets merged, I will either create a follow-up PR or update this one.

if this goes in the right direction, I will then move on implementing 3 (SonarCloud and/or Gcov) and 4 (SonarCloud and/or Clang Static Analyzer).

**Comment by rettichschnidi** [comment]
Some other ideas for later on:
* Code style checks using clang-format (will need some discussion)
* Running binaries with sanitizers (e.g. address, memory, undefined behavior) enabled (maybe helpful for issues like #444)

**Comment by sbernard31** [comment]
@rettichschnidi do you plan to integrate more CI or do you consider this issue is resolved ?

**Comment by sbernard31** [comment]
OK I get my answer,  it seems there are more  ( :point_right:  #567)

**Comment by rettichschnidi** [comment]
Yes, I intend to get another ~3 PRs merged:
* As you have seen, #567, code coverage measurement
* Clang build
* Some kind of static doe analyzis

**Comment by davidgraeff** [comment]
My initial issue is solved. And even adapted, by not using Travis CI which was common in 2016 and suggested by me, but Github actions. The ticket can be closed, thanks :)

**Comment by rettichschnidi** [comment]
From my POV, once PR #570 is merged, the most relevant points are addressed (not perfectly, but good enough for now).

**Comment by rettichschnidi** [comment]
Done! :partying_face:
